microprofile/server/src/main/java/io/helidon/microprofile/server/ServerCdiExtension.java

```diff
@@ -582,4 +594,94 @@ private HttpRouting.Builder findRouting(String className,
 
         return serverNamedRoutingBuilder(routingName);
     }
+
+    // -- CDI bean producers for ServerRequest and ServerResponse -----------------------
+
+    private void afterBeanDiscovery(@Observes AfterBeanDiscovery event) {
+        event.addBean(new ServerRequestProducer());
+        event.addBean(new ServerResponseProducer());
+    }
+
+    /**
+     * Base class for server producers sharing common methods.
+     */
+    static class ServerProducer {
+
+        @SuppressWarnings("all")
+        public static class DefaultAnnotationLiteral extends AnnotationLiteral<Default> implements Default {
+            private static final long serialVersionUID = 1L;
+        }
+
+        public Set<Annotation> getQualifiers() {
+            return Collections.singleton(new DefaultAnnotationLiteral());
+        }
+
+        public Class<? extends Annotation> getScope() {
+            return RequestScoped.class;
+        }
+
+        public Set<Class<? extends Annotation>> getStereotypes() {
+            return Collections.emptySet();
+        }
+
+        public Set<InjectionPoint> getInjectionPoints() {
+            return Collections.emptySet();
+        }
+
+        public boolean isAlternative() {
+            return false;
+        }
+
+        public String getName() {
+            return null;
+        }
+    }
+
+    /**
+     * Producer bean that returns {@code ServerRequest} from active context.
+     */
+    static class ServerRequestProducer extends ServerProducer implements Bean<ServerRequest> {
+
+        @Override
+        public Class<?> getBeanClass() {
+            return ServerRequest.class;
+        }
+
+        @Override
+        public Set<Type> getTypes() {
+            return new HashSet<>(Arrays.asList(ServerRequest.class, HttpRequest.class, Object.class));
```

Comment:
Probably just `Set.of(ServerRequest.class, HttpRequest.class, Object.class)` right? Otherwise this returns a mutable set.